lib/charms/kubernetes_charm_libraries/v0/multus.py

```diff
@@ -537,8 +538,15 @@ def _configure_network_attachment_definitions(self):
             list of NetworkAttachmentDefinitions to create
           - Else, delete it
         2. Goes through the list of NetworkAttachmentDefinitions to create and create them all
+        3. Detects the NAD config changes and triggers pod restart
+           if any there is any modification in existing NADs
         """
         network_attachment_definitions_to_create = self.network_attachment_definitions_func()
+        if network_attachment_definitions_to_create is None:
```

Comment:
This should never be `None`, the type hint of the return value of the passed function is `list[NetworkAttachmentDefinition]`. I am not sure there is any reason to support `None` here.

Reply:
Yes, there is a reason. We are always expecting a return value from network_attachment_definitions_func. However, for the situations that config is not valid, we can not constitute and provide the NetworkAttachmentDefinition using invalid config. For those situation, self.network_attachment_definitions_func returns None and if this is None it will just return without doing anything.  If it returns empty list, this causes to delete all the existing NADs. 
We have another option here.
We can raise a custom Invalid config error but this  puts the charm into error state till we provide a valid config. TBH I  did not like it. Do you have a suggestion ?

Reply:
I modified the type hints.

Reply:
IMO handling this in the charm would be a better solution, especially that with the current solution I don't see any feedback about the invalid config. In the charm, we don't necessarily have to raise an exception; it may be possible (depending on how charm handles events) to put it in a `BlockedStatus`.

Reply:
We are handling it in the charm by putting it in to Blocked status but in the library part some operations (NAD delete/create) are done in the background triggered by config_changed event by using those values. We are trying to prevent those operations for invalid configs.
https://github.com/canonical/sdcore-upf-operator/blob/f5bd36ba780be36c1dba74fc32211fa36cf904d1/src/charm.py#L171

Reply:
OK, in that case I wonder if we need to instantiate `KubernetesMultusCharmLib` in the charm constructor? Alternatively, perhaps it would simply things if we removed events handling from the lib and let the charm do it?

Reply:
That could be another approach to manage the operations from one hand. But it requires to put more on charm part by taking the responsibility from library side. I think, we need more opinions here. @gruyaume @markbeierl

Reply:
I think most of the issues revolve around this lib directly listening on `config_changed` for no good reason (it is not using any configuration directly). This was a design errors in the first implementation I think.

I think the right way to fix this is to provide a method for the charm to trigger reconfiguration. It would let the charm do the expected validation and control the flow of execution.

Reply:
I'm not sure about "the right way to fix this is to provide a method for the charm to trigger reconfiguration". Since this is a charm library, it can simply listen to charm events. Upon instantiation we could simply provide the list of events we want the lib to listen to. In all of our use cases, `config_changed` is the right one as NAD content is derived from Juju configuration values

Reply:
A task is created  to solve this issue independently: https://warthogs.atlassian.net/browse/TELCO-759

Reply:
PR: https://github.com/canonical/kubernetes-charm-libraries/pull/17 is created to solve this problem.

Reply:
The PR is merged and the issue is fixed.